README.md
# [Ad Generator](https://airat.biz/random/)
* Contributors: [Airat Halitov](https://github.com/AiratHalitov)
* Requires at least: 3.8
* Tested up to: 5.4.3
* Stable tag: [2.2.0](https://github.com/AiratHalitov/ad-generator/releases/latest)
* License: GPLv3
* License URI: https://www.gnu.org/licenses/gpl-3.0.html

Профессиональный рандомизатор текста и генератор объявлений от Айрата Халитова / Professional text randomizer and ad generator. Author: Airat Halitov


## Описание

Профессиональный рандомизатор текста и генератор объявлений от Айрата Халитова / Professional text randomizer and ad generator. Author: Airat Halitov

## Установка

1. Visit 'Plugins > Add New'
2. Click 'Upload Plugin'
3. Upload the file 'ad-generator.zip'
4. Activate Ad Generator from your Plugins page.
5. Add [ad_generator] shortcode to WordPress Page.

## Настройка

1. Create new WordPress Page, add [ad_generator] shortcode and save
2. Go to page and use ad generator
3. Enjoy

## Скриншоты
### In Russian:
![ad-generator rus](https://user-images.githubusercontent.com/4050715/44120760-d26f5f34-a036-11e8-839a-da43067ff6a1.png)

### In English:
![ad-generator eng](https://user-images.githubusercontent.com/4050715/44120977-97561f54-a037-11e8-9600-054306c0b98b.png)

### Gif Animation of Process:
![ad-generator animation](https://user-images.githubusercontent.com/4050715/44121069-eda8e5a8-a037-11e8-854d-3d6ac1cb5780.gif)


***

# [Документация](https://github.com/AiratHalitov/ad-generator/wiki)

## Что такое рандомизатор текста?

Это - программа для промышленного создания псевдоуникального контента. Используется, например, при регистрации сайта во множестве каталогов. Чтобы в каждом каталоге сайт описывался уникальным с точки зрения поисковиков текстом. В отличие от аналогичных инструментов (синонимайзеры, доргены) позволяет максимально сохранить читабельность результирующих текстов.

## Как работает рандомизатор?
Берем некий исходный текст. Например: 
 
> Бытует мнение, что и копирайтинг (написание текстов) и рерайтинг (переработка готовых текстов) с успехом можно доверить текстовому рандомизатору – специальной программе.

### Обрабатываем его специальным образом:
1. Если "текст 1" можно заменить на "текст 2" или на "текст 3", то вместо "текст 1" вставляем инструкцию `{текст 1|текст 2|текст 3}`. 
1. Если "текст" можно опустить, то вместо "текст" вставляем инструкцию `{|текст}`. 
1. Если можно перемешать последовательность "текст 1 текст 2 текст 3", то вместо нее вставляем инструкцию `[текст 1|текст 2|текст 3]`. 
1. Если можно перемешать последовательность "текст 1, текст 2, текст 3", то вместо нее вставляем инструкцию `[+,+текст 1|текст 2|текст 3]`. 
1. Если можно перемешать абзацы "абзац1 абзац2 абзац3", то можно просто писать `абзац1|абзац2|абзац3`. 
1. Если нужно вставить в текст какой-то спецсимвол `{, }, |, [, ], + или \`, то его следует экранировать: `\{, \}, \|, \[, \], \+ или \\`.
1. Если нужно вставить случайное число - шаблон `%rand%` возвращает случайное число от 0 до 9

**Инструкции могут иметь неограниченную вложенность.**



### Вариант обработки на скорую руку: 
```
{ {Бытует|Существует|Есть|Имеется} мнение|{Некоторые|Многие} думают}, что и
[+и+
 {копирайтинг (написание текстов)
 |написание текстов (копирайтинг)
 |написание текстов
 |копирайтинг
 }
|{рерайтинг (переработку готовых текстов)
 |переработку готовых текстов (рерайтинг)
 |переработку готовых текстов
 |рерайтинг
 }
] [{|с успехом}|можно] доверить
[+–+{текстовому рандомизатору|рандомизатору {|текста|текстов}}
|{специальной|особой} {программе|проге}
].
```

Скармливаем этот код рандомизатору и получаем кучу вариантов, таких как: 
 
> Некоторые думают, что и написание текстов и рерайтинг (переработку готовых текстов) с успехом можно доверить особой программе – рандомизатору.

или 
 
> Есть мнение, что и переработку готовых текстов и написание текстов можно доверить текстовому рандомизатору – особой проге.

Для приведенного в примере кода существует 24576 различных вариантов результирующих текстов.

### Итак, повторим: 
* `{текст 1|текст 2|текст 3}` - перебор
* `[текст 1|текст 2|текст 3]` - перестановки
* `[+разделитель+текст 1|текст 2|текст 3]` - перестановки с разделителем
* `{ } | [ ] + \` - экранизация спецсимволов в виде `\{, \}, \|, \[, \], \+ или \\`
* `%rand%` возвращает случайное число от 0 до 9
* Поддерживается вложенность дирректив

## Консольная версия скрипта
Данный скрипт теперь можно запускать через коммандную строку, т.е. без установки плагина на WordPress сайт. Это гораздо быстрее и не ограничено ни по ресурсам, ни по выводу результатов. 

Необходимо запустить файл `ad-generator-cli.php`, как обычный php-файл, и на входе в параметрах скрипта задать файл с шаблоном, количеством выводимых вариантов и файл для записи результатов (необязательно):

### Пример команды запуска скрипта:
```
php ad-generator-cli.php -n 300 -f shablon.txt -o result.txt
```
где 

* `-n` или `-N` - количество вариантов на выходе (по умолчанию 300),
* `-f` или `--file` - файл с шаблоном (обязательно),
* `-o` или `--out` - файл с результатом (по умолчанию, `result-N.txt`),
* `-h` или `--help` - вызов справки.

Запускать эту команду нужно из папки плагина. Параметры разделяются пробелом, как в примере. У вас обязательно должны быть права на чтение и запись файлов в этой папке. На выходе вы получите файл, в котором будут сгенерированные по вашему шаблону случайные N вариантов.

## [Примеры шаблонов](https://github.com/AiratHalitov/ad-generator/wiki/%D0%9F%D1%80%D0%B8%D0%BC%D0%B5%D1%80%D1%8B-%D1%88%D0%B0%D0%B1%D0%BB%D0%BE%D0%BD%D0%BE%D0%B2)

### Пример шаблона объявления для ниши «Ремонт квартир»
Попробуйте вставить его в [рандомизатор](https://airat.biz/random/) и посмотрите, что получится на выходе. Это объявления можно публиковать на авито. 

#### Заголовок объявления:
```
Ремонт {квартир|офисов|коттеджей} под ключ. Гарантия, {бригада|качество}
```

#### Текст объявления:
```
Бригада {высококвалифицированных|опытных|ответственных} {ремонтников|работников|мастеров} {выполнит|произведет} {качественный|профессиональный} ремонт {Вашей квартиры|Вашего дома|Вашего жилища} по {разумной|приемлемой} {цене|стоимости}. {У нас|Мы|Ищите у нас|Почему мы|Обращайтесь к нам}:
— {Закупка стройматериалов по {оптовым|сниженным} {ценам|расценкам}|Выполнение {всех видов|любых} работ в {кратчайшие|минимальные} сроки}.
— {{Все|Наши} {рабочие|ремонтники} — граждане РФ|{Опыт|Стаж} {всех|наших} {рабочих|сотрудников|строителей} {5|6|7} лет}.
— Гарантия на {все виды работ|{любые|выполненные} работы} {1|2|3} года.
— {{Все|Любые} виды {ремонтно-отделочных|ремонтных|отделочных} работ {любой сложности|»от А до Я»|под ключ}.|Выезд {специалиста|замерщика} для консультации и замера — БЕСПЛАТНО!}
{===|* * *|***|# # #|~ ~ ~|- — -|___}
{->>|=>|>>|->} ТОЛЬКО до %DATE {действует|работает} СКИДКА на {ремонт|проведение ремонта} {квартир|помещений|Вашей квартиры} «под ключ» – {15|20|10}%!!!
{✆|☏|►} {ЗВОНИТЕ ПО ТЕЛЕФОНУ|ТЕЛЕФОН|ЗВОНИТЕ|Звоните прямо СЕЙЧАС|Звоните и заказывайте БЕСПЛАТНЫЙ замер}: %PHONE
```

---
### Ещё примеры:

#### Дни недели:

```
{Понедельник|Вторник|Среда|Четверг|Пятница|Суббота|Воскресенье}
```

или 

```
{ПН|ВТ|СР|ЧТ|ПТ|СБ|ВС}
```

--- 
#### Месяцы:

```
{Январь|Февраль|Март|Апрель|Май|Июнь|Июль|Август|Сентябрь|Октябрь|Ноябрь|Декабрь}
```

#### Дни месяца:

```
{{|1|2}{0|1|2|3|4|5|6|7|8|9}|30|31}
```

или проще

```
{1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31}
```

#### Числа:

**0-9:**

```
{1|2|3|4|5|6|7|8|9|0}
```

или 

```
%rand%
```

**10-99 (двузначные):**

```
{1|2|3|4|5|6|7|8|9}{1|2|3|4|5|6|7|8|9|0}
```

или
 
```
{1|2|3|4|5|6|7|8|9}%rand%
```

#### Английский алфавит:

```
{A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z}
```

#### Русский алфавит:

```
{А|Б|В|Г|Д|Е|Ё|Ж|З|И|Й|К|Л|М|Н|О|П|Р|С|Т|У|Ф|Х|Ц|Ч|Ш|Щ|Ъ|Ы|Ь|Э|Ю|Я}
```

и т.д. Таким образом можно ставить абсолютно любые конструкции. Всё ограничено только вашей фантазией


***

## [Changelog](https://github.com/AiratHalitov/ad-generator/blob/master/CHANGELOG.md)

### 2.2.0
* Исправлен баг с подменой лишних пробелов
* Рефакторинг кода рандомизатора
* Добавлено больше примеров шаблонов

### 2.0.1
* Исправлен баг с появлением лишних пробелов (#12)

### 2.0.0
* Добавлена новая консольная версия
* Обработка ошибок и справка в консольной версии
* Добавлены теги label к выбору вариантов
* Поправлено удаление лишних пробелов в результатах

### 1.4.0
* Увеличено число допустимых символов с 4000 до 10000
* Исправлен автофокус на странице с формой

### 1.3.2
* Добавлены айдишники к выбору количества результатов
* Обновлен URL лицензии
* Добавлены скриншоты в README.md
* Обновлены файлы для репозитория GitHub (DOCUMENTATION, ISSUE_TEMPLATE, README)
* Изменен регистр названия репозитория Ad-Generator -> ad-generator
* Обновлены файлы переводов

### 1.3.1
* Форматирование вывода результатов
* Обновлены файлы переводов

### 1.3.0
* Добавлена возможность выбора количество отображаемых результатов

### 1.2.3
* `%rand%` возвращает случайное число от 0 до 9
* Исправлена ошибка с лишними символами \
* Форматирование и оптимизация кода

### 1.2.2
* Теперь работает поддержка разных языков (пока русский и английский)

### 1.2.1
* Добавил автоматические обновления через GitHub Updater
* Поправил changelog.md
* Прописал метаданные пакета
* Поправил /languages/
* Поправил шаблон генератора

### 1.2.0
* Добавил поддержку языковых файлов
* Добавил русские и английские переводы
* Обновил "рыбу" для рандомизации
* Уменьшил макс. число символов в форме до 4000

### 1.1.1
* Добавлены айдишники ко всем полям и элементам
* Настроена ширина поля. Оптимизация под мобильники
* Визуальные настройки отображения элементов
* Добавил поддержку composer (файл composer.json)
* Расширил ISSUE_TEMPLATE.md
* Оптимизация кода

### 1.1.0
* Решена проблема с множественными пробелами
* Добавлена ссылка на проект GitHub
* Добавлена кнопка очистки результатов
* Настроено визульное оформление результатов
* Обработка частных случаев разного количества результатов
* Очистка вывода от обилия слешей

### 1.0.2
* Решил проблему с переносом строк
* Добавил файлы для GitHub
* Настроил правильный вывод результатов

### 1.0.1
* Объединил всё в одну функцию. Убрал лишнее и поправил ошибки
* Проверил на тестовом сайте - всё работает
Дальше:
* Нужно навести красоту
* Сделать так, чтобы генерировалось с переносами строк

### 1.0.0
* Initial Release

***

# [Contributing](https://github.com/AiratHalitov/ad-generator/blob/master/CONTRIBUTING.md)
Firstly, thanks for even thinking about contributing. You're awesome!

To make things easier we've created these guidelines:

## Issues:

* When opening an issue please keep it to one bug / enhancement / question etc. this to simplify the discussion.
* Please test the master branch to confirm the issue still exists.
* Be as descriptive as you can. Screen shots are always welcome.

## Pull Requests

* The general rule is to use 1 PR for 1 Issue. This helps the merge master quickly figure out how the new code affects the plugin.
* All Pull Request must be made from int "master". You will be responsible for checking that your branch is up to date.
* All pull requests must be related to an existing / new issue.
* If you have the chops please include Unit Tests along with your Pull Request.

We appreciate all your efforts. Your contributions make Ad-Generator even better!


CHANGELOG.md
## Changelog

### 2.2.0
* Исправлен баг с подменой лишних пробелов
* Рефакторинг кода рандомизатора
* Добавлено больше примеров шаблонов

### 2.0.1
* Исправлен баг с появлением лишних пробелов (#12)

### 2.0.0
* Добавлена новая консольная версия
* Обработка ошибок и справка в консольной версии
* Добавлены теги label к выбору вариантов
* Поправлено удаление лишних пробелов в результатах

### 1.4.0
* Увеличено число допустимых символов с 4000 до 10000
* Исправлен автофокус на странице с формой

### 1.3.2
* Добавлены айдишники к выбору количества результатов
* Обновлен URL лицензии
* Добавлены скриншоты в README.md
* Обновлены файлы для репозитория GitHub (DOCUMENTATION, ISSUE_TEMPLATE, README)
* Изменен регистр названия репозитория Ad-Generator -> ad-generator
* Обновлены файлы переводов

### 1.3.1
* Форматирование вывода результатов
* Обновлены файлы переводов

### 1.3.0
* Добавлена возможность выбора количество отображаемых результатов

### 1.2.3
* %rand% возвращает случайное число от 0 до 9
* Исправлена ошибка с лишними символами \
* Форматирование и оптимизация кода

### 1.2.2
* Теперь работает поддержка разных языков (пока русский и английский)

### 1.2.1
* Добавил автоматические обновления через GitHub Updater
* Поправил changelog.md
* Прописал метаданные пакета
* Поправил /languages/
* Поправил шаблон генератора

### 1.2.0
* Добавил поддержку языковых файлов
* Добавил русские и английские переводы
* Обновил "рыбу" для рандомизации
* Уменьшил макс. число символов в форме до 4000

### 1.1.1
* Добавлены айдишники ко всем полям и элементам
* Настроена ширина поля. Оптимизация под мобильники
* Визуальные настройки отображения элементов
* Добавил поддержку composer (файл composer.json)
* Расширил ISSUE_TEMPLATE.md
* Оптимизация кода

### 1.1.0
* Решена проблема с множественными пробелами
* Добавлена ссылка на проект GitHub
* Добавлена кнопка очистки результатов
* Настроено визульное оформление результатов
* Обработка частных случаев разного количества результатов
* Очистка вывода от обилия слешей

### 1.0.2
* Решил проблему с переносом строк
* Добавил файлы для GitHub
* Настроил правильный вывод результатов

### 1.0.1
* Объединил всё в одну функцию. Убрал лишнее и поправил ошибки
* Проверил на тестовом сайте - всё работает
Дальше:
* Нужно навести красоту
* Сделать так, чтобы генерировалось с переносами строк

### 1.0.0
* Initial Release


example.php
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7; IE=EDGE" />
<meta name="viewport" content="width=1024">

<title>Генератор объявлений</title>

<!-- Bootstrap -->
<link href="http://adgen.besaba.com/css/bootstrap.min.css" rel="stylesheet">


<script src="//vk.com/js/api/openapi.js" type="text/javascript" charset="windows-1251"></script>

<?php
function parse_express($string)
{
  $m =array();
  preg_match_all("/{[a-zA-Zа-яА-ЯЁё0-9\-\+!<>^`~_☎✆☏►>=\?,\/.*;\\\@#\$%\(\)\"\':\|\s]+}/u",$string,$m);
  if (empty($m[0][0]))
  {
      return $string;
  }
  $result = $string;
  foreach ($m[0] as $id =>$val)
  {
     $result = str_replace($val, parse_elem($val),$result);
  }
  return parse_express($result);
}
function parse_elem($string)
{
   $string = preg_replace("/[{}]/","",$string);
   $arr = explode("|",$string);
   $res= ($arr[rand(0,count($arr) -1)]);
   return $res;
}

$a = 'Заголовок (например): Ремонт {квартир|офисов|коттеджей} под ключ. Гарантия, {бригада|качество}
{===|* * *|***|# # #|~ ~ ~|- - -|___}
Бригада {высококвалифицированных|опытных|ответственных} {ремонтников|работников|мастеров} {выполнит|произведет} {качественный|профессиональный} ремонт {Вашей квартиры|Вашего дома|Вашего жилища} по {разумной|приемлемой} {цене|стоимости}. {У нас|Мы|Ищите у нас|Почему мы|Обращайтесь к нам}:
{>|~|+|=>} {Закупка стройматериалов по {оптовым|сниженным} {ценам|расценкам}|Выполнение {всех видов|любых} работ в {кратчайшие|минимальные} сроки}.
{>|~|+|=>} {{Все|Наши} {рабочие|ремонтники} - граждане РФ|{Опыт|Стаж} {всех|наших} {рабочих|сотрудников|строителей} {5|6|7} лет}.
{>|~|+|=>} Гарантия на {все виды работ|{любые|выполненные} работы} {1|2|3} года.
{>|~|+|=>} {{Все|Любые} виды {ремонтно-отделочных|ремонтных|отделочных} работ {любой сложности|"от A до Я"|под ключ}.|Выезд {специалиста|замерщика} для консультации и замера - БЕСПЛAТНO!}
{===|* * *|***|# # #|~ ~ ~|- - -|___}
{->>|=>|>>|->} ТOЛЬКO до %DATE {действует|работает} СКИДКA на {ремонт|проведение ремонта} {квартир|помещений|Вашей квартиры} «под ключ» – {15|20|10}%!!!
{☎|✆|☏|►} {ЗВОНИТЕ ПО ТЕЛЕФОНУ|ТЕЛЕФОН|ЗВОНИТЕ|Звоните прямо СЕЙЧАС|Звоните и заказывайте БЕСПЛАТНЫЙ замер}: %PHONE';
$s = (isset($_POST['text'])) ? $_POST['text']: $a;
?>
</head>

<body>
<table style="width: 900px; margin: 5px auto; border: 0px;">
	<tbody>
	
		<tr>
			<td style="width:620px; margin:5px 5px 5px 5px; vertical-align:top;">


				<form action="/index.php" method="post">
					<textarea class="form-control" rows="20" cols="140" name="text" text="" placeholder="Введите шаблон, например: {Раскрутка|Продвижение|Оптимизация} Ваших {сайтов|ресурсов|порталов} нашими специалистами."><?php echo $s; ?></textarea>
					<br/>
					<p style="text-align:center;"><button class="btn btn-large btn-primary" type="submit"><span style="font-size:32px;">Хочу Oбъявление!</span></button></p>
				</form>
				<br/>
				<?php
				$m=array();
				$a= preg_match_all("/{(([a-zA-Zа-яА-ЯЁё0-9\s]+)\|?)+}/u",$s,$m);
				$aa= (parse_express($s));
				?>
				<p style="text-align: left;">Еще есть <a href="http://adgen.besaba.com/randomizer" target=_blank><strong>бесплатный рандомизатор текста</strong></a></p>

				<p><strong>P.S.</strong> Используйте в шаблоне ТОЛЬКО конструкции с {текс1|текст2}, как показано в примере на этой странице! Допускается любая вложенность.</p>
				<div class="alert alert-info">
				<h4>Готовое объявление для размещения на avito, irr и других ресурсах:</h4>
				<?php echo nl2br($aa); ?>
				</div>


			</td>
		</tr>

	</tbody>
</table>



</body>
</html>


includes/Node.php
<?php

class Node
{
    private $_parent = null;

    private $_str = '';

    private $_type = 'mixing';

    private $_subNodes = array();

    private $_usedKeys = array();

    private $_separator = '';

    private $_isSeparator = false;

    public function __construct(Node $parent = null)
    {
        $this->_parent = $parent;
        if ($parent) {
            $parent->_subNodes[] = $this;
        }
    }

    public function getText()
    {
        $result = '';
        switch ($this->_type) {
            case 'synonyms':
                if (! count($this->_usedKeys)) {
                    $this->_usedKeys = array_keys($this->_subNodes);
                }
                $key = array_rand($this->_usedKeys);
                $result = $this->_subNodes[$key]->getText();
                unset($this->_usedKeys[$key]);
                break;
            case 'mixing':
                shuffle($this->_subNodes);
                foreach ($this->_subNodes as $v) {
                    if ($result) {
                        $result .= '' . $this->_separator;
                    }
                    $result .= '' . $v->getText();
                }
                break;
            case 'series':
                foreach ($this->_subNodes as $v) {
                    $result .= '' . $v->getText();
                }
                break;
            default:
                $result = $this->_str;
        }
        //$result = trim($result);
        //$result = preg_replace('/\s+/u', ' ', $result);
        $result = preg_replace('| +|', ' ', $result);
        /*$result = str_replace(' ,', ',', $result);
        $result = str_replace(' .', '.', $result);
        $result = str_replace(' !', '!', $result);
        $result = str_replace(' ?', '?', $result);
        $result = str_replace('( ', '(', $result);
        $result = str_replace(' )', ')', $result);
        $result = str_replace(' :', ':', $result);*/
        $result = str_replace('%rand%', rand(0, 9), $result);
        return $result;
    }

    public function numVariant()
    {
        $result = 1;
        switch ($this->_type) {
            case 'synonyms':
                $result = 0;
                foreach ($this->_subNodes as $v) {
                    $result += $v->numVariant();
                }
                break;
            case 'mixing':
                for ($i=2; $i<=count($this->_subNodes); ++$i) {
                    $result *= $i;
                }
                foreach ($this->_subNodes as $v) {
                    $result *= $v->numVariant();
                }
                break;
            case 'series':
                foreach ($this->_subNodes as $v) {
                    $result *= $v->numVariant();
                }
                break;
        }
        return $result;
    }

    public function concat($str)
    {
        $str = (string) $str;
        if ($this->_isSeparator) {
            $this->_separator .= $str;
            return $this;
        }
        if ('string' == $this->type) {
            $this->_str .= $str;
            return $this;
        }
        $currentNode = new Node($this);
        $currentNode->setType('string');
        return $currentNode->concat($str);
    }

    public function setType($type)
    {
        switch ((string) $type) {
            case 'string':
                $this->_type = 'string';
                break;
            case 'synonyms':
                $this->_type = 'synonyms';
                break;
            case 'series':
                $this->_type = 'series';
                break;
            default:
                $this->_type = 'mixing';
        }
    }

    public function __get($var) {
        $var = strtolower($var);
        switch ((string) $var) {
            case 'isseparator':
                return $this->_isSeparator;
                break;
            case 'parent':
                return $this->_parent;
                break;
            case 'type':
                return $this->_type;
                break;
            default:
                return null;
        }
    }

    public function __set($var, $value)
    {
        $var = strtolower($var);
        switch ((string) $var) {
            case 'isseparator':
                $this->_isSeparator = (boolean) $value;
        }
    }
}


ad-generator-cli.php
<?php
/**
 * Command line interface for professional text randomizer and ad generator.
 * 
 * USAGE: php ad-generator-cli.php -n 300 -f shablon.txt -o result.txt
 * 
 * -n, -N - number of variants inn output file (default: 300)
 * -f, --file - input file with template (necessary)
 * -o, --out - result file (default: result-N.txt)
 * -h, --help - show help message
 */
/**
 * @package    airathalitov/ad-generator
 * @category   Core
 * @author     Airat Halitov
 * @license    GPLv3
 * @link       https://airat.biz/random/
 * @version    2.2.0
 */

if ( $argc < 2 ) {
    die( show_help () );
}

function show_help () {
echo <<<END
\nHELP:
Version: 2.2.0
Command line interface for professional text randomizer and ad generator.\n
USAGE: php ad-generator-cli.php -n 300 -f shablon.txt -o result.txt\n
Arguments:
  -n, -N \tnumber of variants inn output file (default: 300)
  -f, --file \tinput file with template (necessary)
  -o, --out \tresult file (default: result-N.txt)
  -h, --help \tshow help message

Author: Airat Halitov
Link: https://airat.biz/random/
GitHub: https://github.com/AiratHalitov/ad-generator\n\n
END;
}

function read_file ( $filename ) {
    $fp = fopen( $filename, "r" ) or die( "\nError in read_file(): Unable to open file!\n\n" );
    if ( filesize( $filename ) < 2 ) {
        die( "\nError in read_file(): Input file should not be empty!\n\n" );
    }
    $content = fread( $fp, filesize( $filename ) );
    if ( trim( $content ) == '' ) {
        die( "\nError in read_file(): Input file should not be empty!\n\n" );
    }
    fclose( $fp );
    return $content;
}

function save_file ( $filename, $content ) {
    $fp = fopen( $filename, "w" ) or die( "\nError in save_file(): Unable to open file!\n\n" );
    fwrite( $fp, $content );
    fclose( $fp );
}

$N = 300;
$file_in = '';
$file_out = '';

for( $i = 1; $i < $argc; $i++ )
{
    if ( ( !strcmp( $argv[$i], "-n" ) || !strcmp( $argv[$i], "-N" ) ) && ( $i+1<$argc ) ) {
        $N = ( int )$argv[$i+1];
    }
    if ( ( !strcmp( $argv[$i], "-f" ) || !strcmp( $argv[$i], "--file" ) ) && ( $i+1<$argc ) ) {
        $file_in = ( string )$argv[$i+1];
    }
    if ( ( !strcmp( $argv[$i], "-o" ) || !strcmp( $argv[$i], "--out" ) ) && ( $i+1<$argc ) ) {
        $file_out = ( string )$argv[$i+1]; 
    }
    if ( !strcmp( $argv[$i], "-h" ) || !strcmp( $argv[$i], "--help" ) ) {
        die( show_help () );
    }
}

if ( $N < 1 ) {
    echo "\nWarning! N should be > 0! Using default n = 300\n";
    $N = 300;
}
if ( $N > 1e9 ) {
    echo "\nWarning! N is too big (> 1e9)! Using default n = 300\n";
    $N = 300;
}
if ( $file_in == '' ) {
    echo "\nError! Мissing input file! (argument after -f or --file)\n";
    die( show_help () );
}
if ( $file_out == '' ) {
    $file_out = 'result-' . $N . '.txt';
}

save_file( $file_out, ad_generator_cli( $N, $file_in ) );

function ad_generator_cli( $max_res, $filename ) {
    $result_text = '';
    $ad_text = read_file( $filename );
    $ad_text = str_replace( '\\\\', '\\', $ad_text );
    $ad_text = str_replace( '\\"', '"', $ad_text );
    $ad_text = str_replace( "\\'", "'", $ad_text );

    if ( $ad_text ) {
        require_once 'includes/Randomizer.php';
        $tRand = new Randomizer( $ad_text );
        $num_var = $tRand->numVariant();
        
        if ( $num_var > 1 ) {
            $max_tmp = min( $num_var, $max_res );
            $result_text .= sprintf( "The number of all possible variants: %s. Here are the random %s of them:\n\n", $num_var, $max_tmp );
            
            for ( $i = 0; $i < $max_tmp; ++$i ) {
                $result_text .= $tRand->getText() . "\n\n\n";
            }
        } else {
            $result_text .= "Only 1 possible variant:\n\n";
            $result_text .= $tRand->getText();
        }
    }
    $result_text = preg_replace( "/\n /", "\n", trim( $result_text ) );
    $result_text = preg_replace( "/ \n/", "\n", $result_text );
    return $result_text;
}



ad-generator.php
<?php
/**
 * Plugin Name:        Ad Generator
 * Plugin URI:         https://airat.biz/random/
 * Description:        Professional text randomizer and ad generator.
 * Author:             Airat Halitov
 * Author URI:         https://airat.biz
 * Version:            2.2.0
 * Text Domain:        ad-generator
 * Domain Path:        /languages/
 * Requires WP:        3.8
 * Requires PHP:       5.4
 * Release Asset:      true
 * GitHub Plugin URI:  https://github.com/AiratHalitov/ad-generator
 */
/**
 * @package    airathalitov/ad-generator
 * @category   Core
 * @author     Airat Halitov
 * @license    GPLv3
 * @link       https://airat.biz/random/
 * @version    2.2.0
 */

// Exit if accessed directly
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

class ad_generator_shortcode {
	
	static $max_res = 10;
	static $mydomain = 'ad-generator';
	
	static function init () {
		add_shortcode( 'ad_generator', array( __CLASS__, 'ad_generator_func' ) );
		add_action( 'plugins_loaded', array( __CLASS__, 'ad_generator_textdomain' ) );
	}
	
	
	static function ad_generator_textdomain () {
		load_plugin_textdomain( self::$mydomain, false, basename( dirname( __FILE__ ) ) . '/languages/' );
	}
	
	static function ad_generator_func( $atts ) {
		$result_text = '';
		$result_text .= '<form method="post" action="">';
		$ad_text = isset( $_POST['ad_text'] ) ? (string) $_POST['ad_text'] : '';
		$ad_text = str_replace( '\\\\', '\\', $ad_text );
		$ad_text = str_replace( '\\"', '"', $ad_text );
		$ad_text = str_replace( "\\'", "'", $ad_text );
		
		$result_text .= '<textarea id="ad_text" name="ad_text" cols="80" rows="10" maxlength="10000" style="width: 100%;" placeholder="' . __( 'Введите шаблон', self::$mydomain ) . '">';
		
		if ( $ad_text ) {
			$result_text .= htmlspecialchars( $ad_text );
		} else {
			$result_text .= __( 'Это{|, пожалуй,} самый {лучший|прекрасный|отличный} {рандомизатор|рандомайзер} текста, который я только {видел|встречал}. Он такой [+, +удобный|быстрый] и функциональный{, что ничего другого уже не нужно| - мне всё в нем нравится}{!|.|. : )} {Спасибо!|Спасибо большое!|Спасибо, Айрат!}', self::$mydomain );
		}

		$result_text .= '</textarea><br /><p>';

		$result_text .= __( 'Выбор количества отображаемых результатов (по умолчанию 10):', self::$mydomain );
		$result_text .= '<br /><label><input type="radio" name="ad_count" value="10" id="ads_10" checked> 10</label><br />';
		$result_text .= '<label><input type="radio" name="ad_count" value="100" id="ads_100"> 100</label><br />';
		$result_text .= '<label><input type="radio" name="ad_count" value="300" id="ads_300"> 300</label><br /></p>';
		
		self::$max_res = isset( $_POST['ad_count'] ) ? (int) $_POST['ad_count'] : 10;
		if ( self::$max_res <= 0 ) self::$max_res = 1;
		if ( self::$max_res > 300 ) self::$max_res = 300;

		$result_text .= '<button id="ad_text_btn" class="btn btn-large btn-primary" type="submit">' . __( 'Генерировать', self::$mydomain ) . '</button></form>';
		
		if ( $ad_text ) {
			$result_text .= '<br /><a href=' . $_SERVER['REQUEST_URI'] . ' id="ad_text_clear_btn">' . __( 'Очистить и начать заново', self::$mydomain ) . '</a><br /><br />';
			
			require_once plugin_dir_path( __FILE__ ) . '/includes/Randomizer.php';
			
			$tRand = new Randomizer( $ad_text );
			$num_var = $tRand->numVariant();
			
			if ( $num_var > 1 ) {
				$max_tmp = min( $num_var, self::$max_res );
				$result_text .= sprintf( __( '<p><i>Число всех возможных вариантов: <strong>%s</strong>. Вот случайные <strong>%s</strong> из них:</i></p>', self::$mydomain ), $num_var, $max_tmp );
				
				for ( $i = 0; $i < $max_tmp; ++$i ) {
					$result_text .= '<p id="ad_text_result">' . nl2br( htmlspecialchars( $tRand->getText() ) ) . '</p><hr />';
				}
				if ( $max_tmp >= 10 )
					$result_text .= '<a href=' . $_SERVER['REQUEST_URI'] . ' id="ad_text_clear_btn">' . __( 'Очистить и начать заново', self::$mydomain ) . '</a><br />';
			} else {
				$result_text .= __( '<p><i>Только <strong>1</strong> возможный вариант:</i></p>', self::$mydomain );
				$result_text .= '<p id="ad_text_result">' . nl2br( htmlspecialchars( $tRand->getText() ) ) . '</p><hr />';
			}
		}	
		return $result_text;
	}
}

ad_generator_shortcode::init();
